akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/StashingBehaviors.scala

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright (C) 2017-2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.actor.typed.scaladsl
+
+import java.util.concurrent.TimeoutException
+
+import akka.actor.typed.{ ActorSystem, Behavior }
+import akka.util.{ LineNumbers, PrettyDuration }
+
+import scala.concurrent.{ ExecutionContext, ExecutionContextExecutor, Future }
+import scala.concurrent.duration.FiniteDuration
+import scala.util.{ Failure, Success, Try }
+
+/**
+ * Stashing convenience behaviors that encapsulate common stashing patterns like awaiting a future to complete,
+ * or collecting a number of futures or responses before continuing with further handling of incoming messages.
+ */
+object StashingBehaviors {
+
+  private sealed trait Protocol
+  private final case class FutureCompleted[T](value: Try[T]) extends Protocol
+
+  final class StashFutureFailedException(msg: String, cause: Throwable)
+    extends RuntimeException(msg, cause)
+
+  /**
+   * Keep stashing incoming messages until the passed in [[scala.concurrent.Future]] `f` completes successfully.
+   *
+   * See also [[untilSuccessful]] if you want to react only to the the future's success, and treat failure by
+   * escalating it by failing the actor.
+   */
+  def until[T, M](f: Future[T], buffer: StashBuffer[M], timeout: FiniteDuration)(next: Try[T] ⇒ Behavior[M]): Behavior[M] = {
```

Comment:
I don't think we should mix Future and Actor APIs too much. Stash until a given message predicate `M => Boolean` would be more interesting. Especially since `ctx.ask` is message based and not Future based.

Reply:
If it is _not_ used with ask, I think this is really useful. Having some other API returning a future and wanting to have the actor "wait" for is something I think is quite common in the wild. So providing a version of it that is at least a bit safer would be great.

I also don't think that this introduces any pitfalls that aren't already there if the user needs to do future:y stuff in their actors.

Reply:
I still think the message based should be the core feature, then we could discuss if the Future based removes enough boilerplate to deserve a convenience method.

Without convenience it would look like:
```
StashingBehaviors.until(_ == ExternalTaskDone, buffer, timeout)(next)
implicit val ec = ctx.executionContext
externalTask.onComplete { result =>
  ctx.self ! ExternalTaskDone
}
```